implemented number_clicked to show up on display and clear display

chrisK00 opened the pull request:
Wont display if user tries to start with 0.

chrisK00 committed:
implemented number_clicked to show up on display and clear display

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -19,6 +20,15 @@ public MainPageViewModel()
+        {
+            if (DisplayContent.StartsWith("0"))
+            {
+                DisplayContent = "";
```
Why do you want it to display as empty instead of `0`?

nickmartin1ee7 replied:
Ah, it's so you don't have leading `0`'s in the display. Perhaps we could do something with converting the DisplayContent to a `decimal` data type and converting it back to a `string`. That would remove any leading `0`'s as well!

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -19,6 +20,15 @@ public MainPageViewModel()
 
         #region Methods
         // TODO
```
Remove `// TODO`

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -6,6 +6,7 @@ namespace calculatorUICOOP.ViewModels
     {
+        public string DisplayContent { get; private set; } = "0";
         #region Fields and Properties
         // TODO
```
Remove `// TODO`

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -19,6 +20,15 @@ public MainPageViewModel()
+            }            
+            DisplayContent += textToDisplay;          
+        }
+        public void ClearScreen() { DisplayContent = "0"; }
```
For readability-sake, we should consider spacing this out with newlines

```
public void ClearScreen()
{
    DisplayContent = "0";
}
```

nickmartin1ee7 reviewed (commented):
I see you're linking the display with a public variable in the viewmodel. Perhaps @DonSagiv will have some feedback on how be can change this to a databinding! :)

nickmartin1ee7 reviewed (changes requested):
Try to implement DataContext with the viewmodel and Data bindings to the DisplayLabel

DonSagiv reviewed (changes requested):
Going to agree w/ @nickmartin1ee7 -- Now's a good time as any to implement MVVM. Maybe we should do a coding session so we can work on it together.

chrisK00 committed:
converted string to decimal in order to remove leading zeros

nickmartin1ee7 commented on calculatorUICOOP/Views/MainPage.xaml:
```diff
@@ -11,25 +11,25 @@
             Grid.ColumnSpan="4" HorizontalTextAlignment="Start" VerticalTextAlignment="Center"  FontSize="20">
         </Label>
-        <Label  BackgroundColor="LightYellow" Text="0" x:Name="DisplayLabel" Grid.Row="1" Grid.ColumnSpan="4" 
+        <Label  BackgroundColor="LightYellow" Text="{Binding DisplayContent}" x:Name="DisplayLabel" Grid.Row="1" Grid.ColumnSpan="4"
```
A `Binding` is declared but no DataContext existing linking the viewmodel.
This will require implementing the event as well.
Try reading over this page to get an idea how to do this: https://docs.microsoft.com/en-us/xamarin/xamarin-forms/xaml/xaml-basics/data-bindings-to-mvvm
Reach out to @DonSagiv for some help on this, or a bit of Googling.

nickmartin1ee7 reviewed (changes requested):
The changes you made **do work**; however, I left a comment regarding using Bindings properly since you started to define a binding in the xaml.
Don't be afraid to read documentation or watch videos.
Experiment with new things! :)

chrisK00 committed:
trying mvvm

chrisK00 committed:
not working

chrisK00 committed:
what am i doing

chrisK00 committed:
mvvm

chrisK00 committed:
data binding working

chrisK00 committed:
updated showplusondisplay

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -33,23 +36,23 @@ void OnPropertyChanged([CallerMemberName] string text = null)
+        public void ShowPlusOnDisplay()
         {
-            DisplayContent += textToDisplay;
+            DisplayContent += "+";
```
You are adding a non-number character to a string that is eventually parsed with `decimal.Parse(string)` on line 39.
This will cause a runtime exception (https://i.imgur.com/AWI6XGZ.gif).

nickmartin1ee7 replied:
Instead of hardcoding the `"+"` you could pass in the button text like you where before. This makes the method more functional and able to be reused.

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -33,23 +36,23 @@ void OnPropertyChanged([CallerMemberName] string text = null)
+        public void ShowNumberOnDisplay(string textToDisplay)
         {
             DisplayContent += textToDisplay;
             var removedLeadingZeros = decimal.Parse(DisplayContent);
```
This is unsafely parsing a string.

DonSagiv replied:
Agreed. I'd suggest decimal.TryParse instead.

nickmartin1ee7 commented on calculatorUICOOP/Views/MainPage.xaml.cs:
```diff
@@ -56,10 +55,7 @@ private void Minus_Clicked(object sender, EventArgs e)
-
-         vm.ShowPlusOnDisplay(button.Text);
-     
+            vm.ShowPlusOnDisplay();
```
Perhaps instead of appending the operator on the Displayed text we could store that as the desired operator somewhere in the view model.

nickmartin1ee7 reviewed (changes requested):
**What I need to accept this PR:**
Fix the error regarding the `decimal.Parse()`

**We can open a new issue for the following:**
Look into storing the `decimal` representation of the display. The Displayed Text could be a `ToString()` of that variable. Perhaps it's time to store some data in a new class. We could make a new class that contains the operator and the two numbers being operated on. Think about a math expression, what what class would be able to store something like `4 - 6` ? There are three things in a simple expression we would need to store.

chrisK00 committed:
Update MainPageViewModel.cs

Removed + for now because im unsure what would be the easiest way to handle this+all other op

DonSagiv commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -1,24 +1,58 @@
-    public class MainPageViewModel
+    public class MainPageViewModel : INotifyPropertyChanged
     {
         #region Fields and Properties
```
I like to keep fields and properties in 2 separate regions. They are not fuctionally the same.

DonSagiv commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -1,24 +1,58 @@
+
+        public event PropertyChangedEventHandler PropertyChanged;
+
+        private void OnPropertyChanged([CallerMemberName] string text = null)
```
This looks like a method to me :-) Why is it in "Constructors?"

DonSagiv commented on calculatorUICOOP/Views/MainPage.xaml:
```diff
@@ -2,38 +2,44 @@
+             x:Class="calculatorUICOOP.MainPage"
+            
+          >
+    <ContentPage.BindingContext>
```
Make this "ignorable" (change it to `<d:ContentPage.BindingContext>`). You're instantiating the view-model twice: Once here and the other in the code-behind.

DonSagiv commented on calculatorUICOOP/Views/MainPage.xaml.cs:
```diff
@@ -13,58 +9,65 @@ namespace calculatorUICOOP
     public partial class MainPage : ContentPage
     {
         #region Fields and Properties
```
You have 2 "Fields and Properties" regions. Maybe consolidate them, and then separate the fields from the properties.

DonSagiv commented on calculatorUICOOP/Views/MainPage.xaml.cs:
```diff
@@ -13,58 +9,65 @@ namespace calculatorUICOOP
-        #endregion
 
-        
+        #endregion Constructor
```
The subsequent code looks like a bunch of methods to me, not constructors 😄 
Also it seems like we don't need a majority of these event handlers if we're gonna use Commands.

DonSagiv reviewed (changes requested):
This is a very promising start, but before I can approve this pull request, I'd like to finish the MVVM implementation. I think we should convert the event handlers to Commands, and then handle those accordingly.

DonSagiv committed:
Remove event handlers from code-behind, replace w/ command in ViewModel for number input.

chrisK00 committed:
Input commands for all numbers, operators etc added. DeleteLastInput implemented function. Removed methods in view.cs

chrisK00 committed:
Update MainPageViewModel.cs

Spelling error divide method fixed

DonSagiv commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -27,31 +27,49 @@ public string DisplayContent
+        public ICommand NumberInputCommand { get; set; }
+        public ICommand ClearInputCommand { get; set; }
+        public ICommand DeleteInputCommand { get; set; }
+        public ICommand DivideInputCommand { get; set; }
+        public ICommand MultiplyInputCommand { get; set; }
```
I would make all 4 (add, subtract, multiply, divide) into a single command with a command parameter that has a string. Then use a switch statement to determine the operator from the command parameter and do that operation.

DonSagiv commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -64,10 +82,48 @@ public void ClearScreen()
+        {
+
+
+        }
```
Convert this to a single method with a switch statement. This will make your code less boilerplate.

DonSagiv reviewed (commented):
@nickmartin1ee7 Tell me if you agree with the suggestion I made for the commands.

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -27,31 +27,49 @@ public string DisplayContent
+        public ICommand DeleteInputCommand { get; set; }
+        public ICommand DivideInputCommand { get; set; }
+        public ICommand MultiplyInputCommand { get; set; }
+        public ICommand clearInputCommand { get; set; }
```
Duplicate and unused Command of `ClearInputcommand`

nickmartin1ee7 commented on calculatorUICOOP/ViewModels/MainPageViewModel.cs:
```diff
@@ -64,10 +82,48 @@ public void ClearScreen()
+        {
+
+        }
+        public void ShowdivideOnDisplay(string divide)
```
Duplicate of `ShowDivideOnDisplay`

nickmartin1ee7 reviewed (commented):
I'll go ahead and fix these issues listed in my review.

nickmartin1ee7 committed:
Fixed duplicates

nickmartin1ee7 reviewed (approved):
Verified working. Although the C and CE behavior is a bit unusual at this point in implementation.
